Replace hoedown markdown parser with CMark

redstrate opened the pull request:
This is a draft for patches to replace the old, crusty and buggy hoedown markdown library - with [CMark](https://github.com/commonmark/cmark).

## Benefits
* Less resource management (no more `hoedown.h` needed to wrap api calls)
* Already used in production by other big projects like KDE and GNOME
* Maintained

CMark also supports raw HTML blocks, so stuff like mod pages are no longer broken and have to be worked around (#713):
![image](https://user-images.githubusercontent.com/54911369/211095164-41b294c8-625a-4f2d-98c5-82d00751d5e7.png)

Since we don't need a resource wrapper, and less quirks because hoedown is terrible - it also means less lines of code worrying about markdown rendering :-)

## Tasks
- [x] Check everywhere where markdown is shown and make sure there aren't any rendering problems.
- [x] Vendor/import CMark (see question below)
- [x] Remove hoedown vendored source
- [x] Check around to see if I'm missing any quirks that were previously used with hoedown, I think I zapped all of them though.

## Questions
* Should we vendor in cmark considering it's in a lot of distribution repositories already (unlike hoedown/sundown?)
 It's a small library anyway, so it doesn't matter either way.
* How should we handle the `markdownToHTML` function? I haven't touched the codebase in a while, so I'm not sure if you want anything more than a static global function.

redstrate committed:
refactor: replace hoedown markdown parser with cmark

Signed-off-by: Joshua Goins <josh@redstrate.com>

redstrate committed:
build: remove hoedown vendored source

Signed-off-by: Joshua Goins <josh@redstrate.com>

DioEgizio commented:
> Should we vendor in cmark considering it's in a lot of distribution repositories already (unlike hoedown/sundown?)
It's a small library anyway, so it doesn't matter either way.

Do like zlib/quazip/ghc-filesystem etc and check if it's installed in the system, dynamically link if found and if it isn't found or Launcher_FORCE_BUNDLED_LIBS is on statically link

DioEgizio commented:
Since modrinth uses GitHub flavoured markdown, might make more sense to use cmark-gfm (GitHub's cmark fork) instead 

https://github.com/github/cmark-gfm

DioEgizio commented:
> > Should we vendor in cmark considering it's in a lot of distribution repositories already (unlike hoedown/sundown?)
> > It's a small library anyway, so it doesn't matter either way.
> 
> Do like zlib/quazip/ghc-filesystem etc and check if it's installed in the system, dynamically link if found and if it isn't found or Launcher_FORCE_BUNDLED_LIBS is on statically link

Forgor to say it but cmark-gfm is also commonly packaged (see https://repology.org/project/cmark-gfm/versions ), so yeah do what I said there ^

Scrumplex committed:
refactor: support system and bundled cmark

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
> Should we vendor in cmark considering it's in a lot of distribution repositories already (unlike hoedown/sundown?)
It's a small library anyway, so it doesn't matter either way.

We probably need this at the very least for our MSVC Windows builds.

I have pushed a commit adding support for system installs of CMark as well as falling back to a submodule, if required.

I have also looked at cmark-gfm, but it looks like it doesn't provide CMake files to allow finding it using `find_package`. Might need to use pkgconfig for that.

Scrumplex committed:
fix(nix): add cmark dependency

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

redstrate commented:
> I have also looked at cmark-gfm, but it looks like it doesn't provide CMake files to allow finding it using `find_package`. Might need to use pkgconfig for that.

Erk, that's annoying :-) But if we want to switch to GFM for whatever reason - it shouldn't change anything API wise so that's something we can put off doing until later. It's not like I encountered anything in the changelogs or mod pages I've viewed that requires GH extensions

DioEgizio commented:
I guess not having gfm extensions is not a big issue for now, but flatpak build for some reason is failing

redstrate commented:
> I guess not having gfm extensions is not a big issue for now, but flatpak build for some reason is failing

The fix should be simple, I just need to add cmark to the flatpak manifest. I'll be adding a commit to fix that :-)

redstrate committed:
fix(flatpak): enable builddir

Signed-off-by: Joshua Goins <josh@redstrate.com>

redstrate commented:
Also, is there a standard format for writing new license headers? I see a mix of some of the old license headers from the MultiMC/PolyMC days but also SPDX. I'm guessing I should mark it GPLv3 with SPDX.

DioEgizio commented on launcher/Markdown.h:
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PrismLauncher - Minecraft Launcher
```
```suggestion
 *  Prism Launcher - Minecraft Launcher
```

redstrate replied:
Also fixed!

DioEgizio commented on launcher/Markdown.h:
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PrismLauncher - Minecraft Launcher
+ *  Copyright (C) 2022 Joshua Goins <josh@redstrate.com>
```
```suggestion
 *  Copyright (C) 2023 Joshua Goins <josh@redstrate.com>
```

redstrate replied:
Fixed, oops :-)

redstrate committed:
fix: Remove extra line breaks for modrinth descriptions

Signed-off-by: Joshua Goins <josh@redstrate.com>

redstrate commented:
I also added an additional fix, I noticed that Modrinth pages specifically has extra line breaks. These are now stripped (Curseforge and other mod platforms I tested don't suffer from this issue):

![image](https://user-images.githubusercontent.com/54911369/211169587-66b9b34a-4205-4157-a85a-700bb9201368.png)

How it looked before:

![image](https://user-images.githubusercontent.com/54911369/211169589-2d17c904-e023-4bed-9cc0-660e8ba9d93e.png)

redstrate commented:
I also tested every place I could (with the exception of the updater, I don't know how to test that easily). The commits are all signed, messages fixed up and ready for review now :-)

DioEgizio commented:
just found a mod using gfm stuff

https://modrinth.com/mod/immediatelyfast

uses tables

<img width="346" alt="image" src="https://user-images.githubusercontent.com/83089242/211187895-7d1b906d-662a-4d03-83f5-e8012c5fce0f.png">

redstrate commented:
> just found a mod using gfm stuff

Erk :-) I'll look into changing it to the GFM fork then

redstrate commented:
I got around to working on this again, but looking through cmark gfm is nothing but pain.

Not only are they a fork of cmark - that part is fine - but they don't include important commits like https://github.com/github/cmark-gfm/commit/c8d5be98336cfbeb833938077ebaef6af4e11d8a which fixes CMake `find_package` as shown above. I could make a PR upstream to fix that though.

Also the availability of [cmark-gfm pales](https://repology.org/project/cmark-gfm/versions) in comparison to [regular cmark](https://repology.org/project/cmark/versions). Not that this is a problem if we include it as a submodule, this only matters for linux packaging.

And to include the relevant GitHub extensions (notably tables) I would have to expand the `markdownToHtml` function because there's no simple way to add extensions through the cmark API. Not that big of a deal, but annoying.

Okay, so let's just accept all of those problems and just take the plunge:

```
/usr/include/cmark-gfm-core-extensions.h:10:10: fatal error: config.h: No such file or directory
   10 | #include "config.h" // for bool
      |          ^~~~~~~~~~
compilation terminated.
```

No way, it includes a [private header](https://github.com/github/cmark-gfm/issues/244) in a public API, so even if I wanted to include cmark-gfm support I couldn't :-( Oh there's an [open PR for that](https://github.com/github/cmark-gfm/pull/248) but it hasn't been merged yet. Since there's been no activity, Nix and Ghostwriter had to implement workarounds for this bug, [Nix PR here](https://github.com/NixOS/nixpkgs/pull/162118) and [ghostwriter commit here](https://github.com/synkathairo/ghostwriter/commit/c3b2dc12f6b4193fcf752a6130c871872d19965d) respectively.

Is there that many mods that use tables in their READMEs? It's a shame that they don't render, but honestly cmark-gfm is a baffling library. We could vendor cmark-gfm and then apply our own patches, but I think that's a bad idea for a lot of reasons (and ideally, this would be fixed upstream).

I think this PR is in a good state right now and we can look at cmark-gfm support later, because with just plain cmark our markdown is much better than before :-) Perfection cannot be the enemy of good, etc

DioEgizio commented:
Yeah let's just use normal cmark then

Scrumplex commented on COPYING.md:
```diff
@@ -156,24 +156,6 @@
     the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
     Boston, MA 02110-1301, USA.
 
-## Hoedown
-
-    Copyright (c) 2008, Natacha Porté
-    Copyright (c) 2011, Vicent Martí
-    Copyright (c) 2014, Xavier Mendez, Devin Torres and the Hoedown authors
-
-    Permission to use, copy, modify, and distribute this software for any
-    purpose with or without fee is hereby granted, provided that the above
-    copyright notice and this permission notice appear in all copies.
-
-    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
-    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
-    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
-    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
-    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
-    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
-    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
-
```
needs cmark copyright notice.

Probably just the first section in https://github.com/commonmark/cmark/blob/master/COPYING

redstrate replied:
Oops, forgot about that file :-) Added

Scrumplex commented on flatpak/org.prismlauncher.PrismLauncher.yml:
```diff
@@ -27,6 +27,16 @@ finish-args:
   - --filesystem=xdg-download:ro
 
 modules:
+  - name: cmark
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DCMARK_TESTS=OFF
+      - -DCMARK_BUILD_TYPE=Release
+      - -DCMAKE_INSTALL_PREFIX=/app
+    sources:
+      - type: git
+        url: https://github.com/commonmark/cmark.git
+    builddir: true
```
instead of this, it might be sufficient to enable `builddir` for the main `prismlauncher` module

redstrate replied:
Good idea, changed that too

DioEgizio committed:
fix: add cmark:p to mingw build 

this way we can just dynamically link it on that build instead of building it ourselves and statically linking it

Signed-off-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>

flowln commented on libraries/README.md:
```diff
@@ -18,12 +18,6 @@ See [github repo](https://github.com/FeralInteractive/gamemode).
 
 BSD-3-Clause licensed
 
-## hoedown
```
shouldn't this be replaced with cmark instead of completely removed?

redstrate replied:
Oh I forgot about that file too, changing :-)

flowln reviewed (approved):
lgtm, thanks!

DioEgizio commented:
When you rebase develop with signoff:

redstrate commented:
Oops, somehow a bunch of unrelated commits made it's way into here :S
